Reduce garbage generation in `prometheus_text_format`

the-mikedavis opened the pull request:
This is a somewhat small set of patches to `prometheus_text_format` which aim to reduce garbage creation during registry formatting. Reducing garbage creation drives down the cost to the VM of scraping large registries - both in terms of peak memory allocation and also the work that the garbage collector must do.

With these changes I see reduction in allocation reported by `tprof` in a stress test of one of RabbitMQ's most expensive registries. In a test against single-instance RabbitMQ brokers on EC2 instances this saves a noticeable amount of peak memory and reduces CPU utilization significantly.

<details><summary>tprof testing instructions</summary>

1. Clone `https://github.com/rabbitmq/rabbitmq-server`
2. `cd rabbitmq-server`
3. `make deps`
4. `make run-broker`
5. In another terminal in the `rabbitmq-server` repo, `sbin/rabbitmqctl import_definitions path/to/100k-classic-queues.json` pointing to [this definitions file](https://github.com/rabbitmq/sample-configs/blob/main/queues/100k-classic-queues.json).
6. In the shell from the `make run-broker` terminal, start `tprof` tracing for new processes: `tprof:start(#{type => call_memory}), tprof:enable_trace(new), tprof:set_pattern('_', '_', '_').`
7. In another terminal scrape the expensive endpoint: `curl -v localhost:15692/metrics/per-object --output /dev/null`
8. When that's done, collect and format the sample: `tprof:format(tprof:inspect(tprof:collect())).`

To test this change, <kbd>Ctrl</kbd><kbd>c</kbd> twice out of `make broker`, `cd deps/prometheus` and check out this branch. Then `rm -rf ebin` in that directory, `cd ../../` and repeat steps 4, 6, 7 and 8 again (skipping definitions import).

---

</details>

<details><summary>Registry collection tprof measurement before this change...</summary>

```
****** Process <0.301089.0>  --  100.00% of total *** 
FUNCTION                                                                                   CALLS      WORDS    PER CALL  [    %]
... removed everything less than 1% ...
prometheus_text_format:render_labels/1                                                   2308195    1944642        0.84  [ 1.01]
erlang:atom_to_binary/2                                                                   651584    2375647        3.65  [ 1.23]
prometheus_rabbitmq_core_metrics_collector:'-emit_queue_info/3-fun-0-'/3                  100000    2500000       25.00  [ 1.29]
prometheus_model_helpers:counter_metric/2                                                 301325    3615900       12.00  [ 1.87]
prometheus_text_format:'-render_labels/1-fun-0-'/2                                        321434    4178642       13.00  [ 2.16]
prometheus_rabbitmq_core_metrics_collector:'-collect_metrics/2-lc$^1/1-0-'/2             2300145    4400076        1.91  [ 2.28]
prometheus_model_helpers:'-metrics_from_tuples/2-lc$^0/1-0-'/2                           2308456    4616300        2.00  [ 2.39]
lists:'-filter/2-lc$^0/1-0-'/2                                                           2408461    4816304        2.00  [ 2.49]
erlang:integer_to_binary/1                                                               2206892    6620701        3.00  [ 3.43]
prometheus_rabbitmq_core_metrics_collector:label/1                                       2200038   11000022        5.00  [ 5.69]
prometheus_rabbitmq_core_metrics_collector:'-collect_metrics/2-lc$^0/1-1-'/2             2300145   11500190        5.00  [ 5.95]
prometheus_text_format:'-emit_mf_metrics/2-fun-0-'/3                                     2308150   11541419        5.00  [ 5.97]
prometheus_model_helpers:gauge_metric/2                                                  2006812   24081744       12.00  [12.47]
prometheus_text_format:has_special_char/1                                               23475329   24147190        1.03  [12.50]
prometheus_text_format:render_series/3                                                   2308200   32511401       14.09  [16.83]
ets:match_object/2                                                                            19   38406095  2021373.42  [19.88]
                                                                                                  193184463              [100.0]
```

---

</details>

<details><summary>Registry collection tprof measurement after this change...</summary>

```
****** Process <0.401000.0>  --  99.99% of total *** 
FUNCTION                                                                                  CALLS      WORDS    PER CALL  [    %]
... removed everything less than 1% ...
prometheus_model_helpers:label_pair/1                                                    429393    1717572        4.00  [ 1.16]
prometheus_text_format:render_labels/1                                                  2308195    1944642        0.84  [ 1.32]
erlang:atom_to_binary/2                                                                  651584    2375647        3.65  [ 1.61]
prometheus_rabbitmq_core_metrics_collector:'-emit_queue_info/3-fun-0-'/3                 100000    2500000       25.00  [ 1.69]
prometheus_model_helpers:counter_metric/2                                                301325    3615900       12.00  [ 2.45]
prometheus_text_format:'-render_labels/1-fun-0-'/2                                       321434    4178642       13.00  [ 2.83]
prometheus_rabbitmq_core_metrics_collector:'-collect_metrics/2-lc$^1/1-0-'/2            2300145    4400076        1.91  [ 2.98]
prometheus_model_helpers:'-metrics_from_tuples/2-lc$^0/1-0-'/2                          2308456    4616300        2.00  [ 3.13]
lists:'-filter/2-lc$^0/1-0-'/2                                                          2408461    4816304        2.00  [ 3.26]
erlang:integer_to_binary/1                                                              2206892    6620705        3.00  [ 4.49]
prometheus_rabbitmq_core_metrics_collector:label/1                                      2200038   11000022        5.00  [ 7.45]
prometheus_rabbitmq_core_metrics_collector:'-collect_metrics/2-lc$^0/1-1-'/2            2300145   11500190        5.00  [ 7.79]
prometheus_text_format:render_series/4                                                  2308200   11541000        5.00  [ 7.82]
prometheus_text_format:render_value/2                                                   2308200   11543618        5.00  [ 7.82]
prometheus_model_helpers:gauge_metric/2                                                 2006812   24081744       12.00  [16.32]
ets:match_object/2                                                                           19   38406095  2021373.42  [26.02]
                                                                                                 147597866              [100.0]
```

---

</details>

So with this change, the Cowboy request process in charge of this endpoint allocates `147_597_866` words instead of `193_184_463`, a reduction of `45_586_597` words or 23.6%.

<details><summary>Stress-testing on EC2...</summary>

On EC2 I have two `m7g.xlarge` instances running RabbitMQ: `galactica` which carries this change and `kestrel` which uses `prometheus` at v5.1.1 (latest version RabbitMQ has adopted). A third instance `curl`s these instances at an interval of two seconds with this script:

```sh
#! /usr/bin/env bash

N=600
SLEEP=2
for i in $(seq 1 $N)
do
  echo "Sleeping ${SLEEP}s... ($i / $N)"
  sleep $SLEEP
  echo "Ask for metrics from $1... ($i / $N)"
  curl -s "http://$1:15692/metrics/per-object" --output /dev/null &
done

wait
```

This asynchronously fires off a scrape request every two seconds for twenty minutes. The third node runs this script against both `galactica` and `kestrel` at the same time. The third node also scrapes these nodes' `node_exporter` metrics and RabbitMQ prometheus endpoint for Erlang allocator metrics.

### `kestrel` (baseline)

##### Instance-wide memory usage

<img width="1628" height="565" alt="grafana-kestrel-mem" src="https://github.com/user-attachments/assets/7a8d91d7-f958-42ba-a489-baa17b5fcdb1" />

##### Instance-wide CPU usage

<img width="1619" height="564" alt="grafana-kestrel-cpu" src="https://github.com/user-attachments/assets/166ecdc7-bb33-42ae-aebf-0d1ed8f81cd0" />

##### Erlang allocators

<img width="1629" height="611" alt="grafana-kestrel-erlang-alloc" src="https://github.com/user-attachments/assets/02dc33d0-022f-4054-b76a-73c25a249448" />

### `galactica` (this branch)

##### Instance-wide memory usage

<img width="1621" height="565" alt="grafana-galactica-mem" src="https://github.com/user-attachments/assets/c0f3d513-3195-447f-84d2-eb81efa00170" />

##### Instance-wide CPU usage

<img width="1622" height="566" alt="grafana-galactica-cpu" src="https://github.com/user-attachments/assets/7fa14f71-daa6-4cc7-bf34-10f903313761" />

##### Erlang allocators

<img width="1618" height="611" alt="grafana-galactica-erlang-alloc" src="https://github.com/user-attachments/assets/eacf2dee-6c85-46c2-b632-d8677540f1a0" />

---

</details>

We can see `kestrel` (baseline) pinned at around 95% CPU usage consistently, hovering at around 9-10 GB instance-wide memory usage and the VM aware of 3.5-4.5 GB of usage. And `galactica` (this branch) sitting at 50% CPU usage, around 7.5-8.5 GB instance-wide memory and the VM tracking around 2-3 GB of memory.

While the peak memory usage is reduced nicely, the main benefit is the CPU is loaded much less than before - I assume from performing less garbage collection.

the-mikedavis committed:
prometheus_text_format: Avoid allocation in `has_special_char/1`

`prometheus_text_format:has_special_char/1` is called very often when
a registry contains many metrics with label pairs. We can use
`binary:match/2` to search within a label binary for the special
characters (newline, backslash and double-quote) without allocation.

The old code using binary match syntax creates a match context every
time the function is called (except, not recursion - then the match
context is reused). A match context allocates 5 words to the process
heap when it is created. When matching many many binaries this scales to
create a noticeable amount of short-lived garbage.

In comparison `binary:match/2` with a precompiled match pattern does not
allocate. The BIF for it is also very well optimized, using `memchr`
since OTP 22.

NelsonVides commented:
This looks amazing! I see it is still marked as draft so I guess no rush, and I'm away from the computer for a few days so only having a look at this on my phone now. Nevertheless, I'd love to see this ready and merged, thank you so much for the work 😃

the-mikedavis commented:
Yep no real rush on this! Looks like I have some work to do to make the CI happy anyways

lhoguin commented:
I've approved the workflow run.

NelsonVides reviewed (commented):
I'm still away from a proper computer so still quickly looking at this from my phone. Thanks a lot for having it pass CI!

I have a question, for good-looking code, the process dictionary's a bit ugly. Is there a way to refactor those usages to pass something more functional or does it have a performance impact that way?

Also for the compiled regex, perhaps storing it in a persistent term that is created at startup could help that perform even faster? It would literally be compiled only once through the entire VM lifetime and never require any GC.

If you don't see an easy way to improve that then it's probably fine to merge this way and later when I'm back to a full computer I'll try to refactor and tag you in a potential PR 🤔

the-mikedavis commented:
Yeah the process dictionary part here is really icky. I thought in my earlier testing that I saw `persistent_term:get/2` allocating but, trying again, looks like I was wrong. So the `binary:match/2` pattern we should definitely move into a persistent_term.

For the `erase/1`+`put/2` dance in `format_into/3` we can't be more functional because the `Collector:collect_mf/2` callback returns `ok`, so we don't have a nice way to accumulate values. To fix that we'd need a really big breaking change :/

NelsonVides commented on src/formats/prometheus_text_format.erl:
```diff
@@ -84,51 +114,18 @@ escape_label_value(LValue) when is_list(LValue) ->
+render_metrics(_Name, [], Bytes) ->
+    Bytes;
+render_metrics(Name, [Metric | Rest], Bytes) ->
+    render_metrics(Name, Rest, render_metric(Bytes, Name, Metric)).
```
I like how this becomes tail-recursive. Note nevertheless, that if we want to be really mad about performance, the shuffling in the params ordering here is not ideal. The BEAM assigns each function input to X registers, so before calling `render_metric(Bytes, Name, Metric)`, it would have to shuffle X0 to X1, and X2 to X0. The JIT will turn this into CPU registers so it is insanely fast, but it will not reorder them into a faster version because it would mess up with tracing. So instead we could be calling here `render_metric(Name, Metric, Bytes)` and the only argument arrangement would be by extracting `Metric` into X1. Likely to be in the order of 3 CPU cycles the whole thing, but `gen_server` code takes great care to do this right so, why not 😄

NelsonVides commented on src/formats/prometheus_text_format.erl:
```diff
@@ -57,16 +57,46 @@ format() ->
+            put(?MODULE, Acc),
+            Callback = fun(
+                #'MetricFamily'{name = Name0, help = Help, type = Type, metric = Metrics}
+            ) ->
```
I tend to do tracing often for debugging runtime systems and anonymous functions are then one of the most unreadable parts of the whole experience. They will print as something like `format_info-1-3942/2` `format_info-1-3294/2`, which is quite unreadable to figure out which is which and what does it actually do. Unfortunately here the anonymous functions are capturing environment so we can't just create a plain "`format_info_collect_mf_callback/2`", but we could have a "`format_info_collect_mf_callback/3`" that actually takes the variables we're going to capture and returns the anonymous function as-is. Then tracing will show "`format_info_collect_mf_callback-1-4324/2`" which ain't perfect but at least has a unique and expressive enough name. Same idea for the other anonymous function in this part.

NelsonVides commented on src/formats/prometheus_text_format.erl:
```diff
@@ -57,16 +57,46 @@ format() ->
-    {ok, Str} = file:pread(Fd, 0, Size),
-    ok = file:close(Fd),
-    Str.
+    format_into(Registry, <<>>, fun(Data, Acc) -> <<Acc/binary, Data/binary>> end).
```
As in the other comment about anonymous functions but this one is easy:
```suggestion
    format_into(Registry, <<>>, fun format_into_acc/2).

format_into_acc(Data, Acc) ->
    <<Acc/binary, Data/binary>>.
```

Provided any better naming you could think of. Also, why not swapping the param order? Isn't it more readable 

```erl
format_into_acc(Acc, Data) ->
    <<Acc/binary, Data/binary>>.
```

And likewise somewhere below

```erl
    Fmt(State1, <<"\n">>).
```

That way it is more intuitive that what we're probably doing is appending.

NelsonVides reviewed (commented):
There's so many checks against whether something is a binary and if it is not then `iolist_to_binary`, that I really wish everything had been just binaries to begin with. But that's a breaking change for another day 😄

Anyway, I have a couple of comments regarding a bit more crazy performance optimisations and making tracing more readable. It's not really important and if it is too annoying I could just shuffle this code myself in a couple of weeks. But maybe you like the idea or you didn't know those tricks and want to do it yourself so just sharing. WDYT? 🙂

the-mikedavis committed:
prometheus_text_format: Optimize binary construction

The formatting callback for a registry can build each metrics family as
a single binary in order to reduce garbage. This mainly involves passing
the accumulator binary through all functions that append to it.

It's more efficient to append to the resulting binary than to allocate
smaller binaries and then append them. For example:

    <<Blob/binary, Name/binary, "_", Suffix/binary>>.
    %% versus
    Combined = <<Name/binary, "_", Suffix/binary>>,
    <<Blob/binary, Combined/binary>>.

The first expression generates less garbage than the second. A good
example of this was the `add_brackets/1` function which was inlined.
Inlining does not turn the first expression (above) into the second
according to the compiler unfortunately, so we pay the cost of creating
a binary with brackets and then formatting that into the larger blob,
rather than formatting in just by copying. This change manually inlines
`add_brackets/1` into its caller `render_series/4`.

This change also changes some list strings into binaries. Especially for
ASCII, strings binaries are _far_ more compact than lists. Lists need
two words per ASCII character - one for the character and one for the
tail pointer. So it's like UTF-32 but worse, basically UTF-128 on a 64
bit machine. ASCII or UTF-8 text in binaries takes one byte per
character in the binary's array, plus a word or two of metadata. E.g.
`<<"hello">>` allocates three words while `"hello"` allocates ten.

the-mikedavis committed:
prometheus_text_format: Eliminate `ram_file`, add `format_into/3`

Building on the work in the parent commit, now that the data being
passed to the `ram_file` is a binary, we can instead build the entire
output gradually within the process. We pay in terms of I/O overhead
from writing and then reading from the `ram_file` since `ram_file` is a
port - all data is passed between the VM and the port driver. The memory
consumed by a port driver is also invisible to the VM's allocator, so
large port driver resource usage should be avoided where possible.

Instead this change refactors the `registry_collect_callback` to fold
over collectors and build an accumulator. The `create_mf` callback's
return of `ok` forces us to store this rather than pass and return it.
So it's a little less hygienic but is more efficient than passing data
in/out of a port.

This also introduces a function `format_into/3` which can use this
folding function directly. This can be used to avoid collecting the
entire response in one binary. Instead the response can be streamed
with `cowboy_req:stream_body/3` for example.

the-mikedavis commented:
All of that sounds good to me - I applied all of the suggestions. For the `render_metrics/3` one I tweaked it slightly so that the binary is always the first arg

NelsonVides commented on src/formats/prometheus_text_format.erl:
```diff
@@ -233,29 +239,17 @@ render_label_pair(B) when is_binary(B) ->
+render_value(Bytes, Value) when is_integer(Value) ->
+    <<Bytes/binary, (integer_to_binary(Value))/binary, "\n">>;
+render_value(Bytes, Value) ->
+    <<Bytes/binary, (iolist_to_binary(io_lib:format("~p", [Value])))/binary, "\n">>.
```
Is there any scenario why this wouldn't be a float? IIRC all values will either be undefined, an integer, or a float. If that assumption is correct, and considering the prometheus protocol actually loves floats, `float_to_binary/1-2` would be much faster and allocate no intermediate garbage here 🤔

the-mikedavis replied:
Looks like `'Gauge'`, `'Counter'`, `'Untyped'` etc allow for a few other values (looking at `prometheus_model.hrl`). We could add branches for those too though - avoiding `io_lib:format/2` would most likely be a nice improvement. Or even just adding a branch for floats.

NelsonVides replied:
Well yeah, floats will very likely be a hot path, other values would rather not happen much in reality.

Checking the docs, these seem like the possible values: https://prometheus.io/docs/specs/om/open_metrics_spec_2_0/#values, where numbers are described at: https://prometheus.io/docs/specs/om/open_metrics_spec_2_0/#numbers so whatever is the right way to print floats using https://www.erlang.org/doc/apps/erts/erlang.html#float_to_binary/2 will surely help performance here.

the-mikedavis replied:
Looks like this is how it's currently done: https://github.com/prometheus-erl/prometheus.erl/blob/a4f14320781787f838b2ddf8f886cd02c279bee7/src/formats/prometheus_text_format.erl#L276

That should probably also use binaries instead.

the-mikedavis replied:
To match `io_lib:format/2` I added a branch for floats that uses `float_to_binary(Value, [short])`, and a TODO comment about it. Maybe we should update that `bound_to_label_value/1` function to use `[short]` instead. But I'd like to avoid any possibly breaking changes with this PR on the other hand.

We still make some garbage for this with `float_to_binary/2` but avoiding all of that allocation in `io_lib:format/2` is a nice win:

```
1> tprof:profile(erlang, float_to_binary, [1520879607.789], #{type => call_memory}).

****** Process <0.87.0>  --  100.00% of total *** 
FUNCTION                  CALLS  WORDS  PER CALL  [     %]
erlang:float_to_binary/1      1      6      6.00  [100.00]
                                     6            [ 100.0]
ok
2> tprof:profile(io_lib, format, ["~p", [1520879607.789]], #{type => call_memory}).

****** Process <0.93.0>  --  100.00% of total *** 
FUNCTION                       CALLS  WORDS  PER CALL  [    %]
io_lib_format:collect/2            2      2      1.00  [ 1.38]
io_lib_format:build_small/1        2      2      1.00  [ 1.38]
io_lib_format:build_limited/5      2      2      1.00  [ 1.38]
io_lib_format:modifiers/3          1      4      4.00  [ 2.76]
io_lib_format:precision/2          1      4      4.00  [ 2.76]
io_lib_format:field_value/2        1      4      4.00  [ 2.76]
io_lib_format:pad_char/2           1      4      4.00  [ 2.76]
io_lib_format:field_width/3        1      5      5.00  [ 3.45]
io_lib_format:collect_cc/2         1      7      7.00  [ 4.83]
io_lib_format:count_small/2        2     12      6.00  [ 8.28]
erlang:float_to_list/2             1     28     28.00  [19.31]
io_lib_format:print/11             3     35     11.67  [24.14]
io_lib_format:collect_cseq/2       1     36     36.00  [24.83]
                                        145            [100.0]
ok
```

NelsonVides replied:
Right, that's all wonderful findings. Thank you so much for this change!

NelsonVides reviewed (commented):
Loving it. Got one more question :D

NelsonVides commented on src/formats/prometheus_text_format.erl:
```diff
@@ -172,33 +175,36 @@ render_metric(Name, #'Metric'{
-    Bytes2 = <<Bytes1/binary, (render_series([Name, "_count"], LString, Count))/binary>>,
-    Bytes3 = <<Bytes2/binary, (render_series([Name, "_sum"], LString, Sum))/binary>>,
+    Bytes2 = render_series(Bytes1, <<Name/binary, "_count">>, LString, Count),
+    Bytes3 = render_series(Bytes2, <<Name/binary, "_sum">>, LString, Sum),
```
Also this could probably just return the result without binding the `Bytes3` variable 🤔

NelsonVides commented on src/formats/prometheus_text_format.erl:
```diff
@@ -137,12 +140,13 @@ render_metric(Name, #'Metric'{
-    Bytes1 = render_series([Name, "_count"], LString, Count),
-    Bytes2 = <<Bytes1/binary, (render_series([Name, "_sum"], LString, Sum))/binary>>,
+    Bytes1 = render_series(Bytes0, <<Name/binary, "_count">>, LString, Count),
+    Bytes2 = render_series(Bytes1, <<Name/binary, "_sum">>, LString, Sum),
```
Likewise at the end of the fold below, no need to bind Bytes3.

NelsonVides reviewed (commented):
Just dirty stuff, this was there before your code changes anyway, it just now shows on the diff.

NelsonVides reviewed (approved):
A blast of a change! Thank you so much for all the effort and also, for keeping such a tidy git history :)

NelsonVides commented:
@the-mikedavis gonna get a release to hex done over the weekend 👍🏽

the-mikedavis commented:
Sweet, thanks @NelsonVides!

NelsonVides commented:
Published https://hex.pm/packages/prometheus/6.1.0 🎉
